### ISSUE TITLE
Tab-line-mode: not possible to switch to unreal buffers

### ISSUE THREAD

**Issue by tbruckmaier** [hlissner/doom-emacs#3495]
**What did you expect to happen?**
When `tab-line-mode` is active and a `*scratch*` or any unreal buffer is shown as a tab, clicking it should select that buffer.

**What actually happened?**
Another buffer is selected instead (a "real" one in doom terms)

**Additional details:**
This behaviour can be seen in the doom sandbox: `M-x doom/sandbox`, `C-c C-p` to open a new emacs instance without any custom config.

In the sandbox window: open any file via `M-x find-file` and enable tab-line-mode via `M-x global-tab-line-mode`. Now there are three tabs: `*scratch*`, `*doom*` and the opened file. Selecting the buffers by clicking the according tabs works fine with `*doom*` and the file buffer, but not with `*scratch*` (or more generally: it works with "real" buffers, but not with "unreal" ones). If `*scratch*` is marked as real (via `(doom-mark-buffer-as-real-h)` as described in #2891), switching to it with a click on the tab works as expected.

Apparently it is intended, that several functions like `switch-to-next-buffer` skip unreal buffers, but with the tab-line the behaviour seems unexpected? A click on a tab should select the according buffer right? Or should unreal buffers not be displayed in the tab-line in the first place? `tab-line-tabs-function` is `tab-line-tabs-window-buffers` (the default value)

**System information:**
<details><pre>

```
emacs   version    27.0.91
        features   XPM JPEG TIFF GIF PNG RSVG CAIRO SOUND GPM DBUS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE HARFBUZZ LIBOTF ZLIB TOOLKIT_SCROLL_BARS GTK3 X11 XDBE XIM MODULES THREADS LIBSYSTEMD JSON PDUMPER LCMS2 GMP
        build      Jun 08, 2020
        buildopts  (--prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --localstatedir=/var --mandir=/usr/share/man --with-gameuser=:games --with-sound=alsa --with-modules --without-gconf --without-gsettings --with-x-toolkit=gtk3 --without-xaw3d --without-m17n-flt --with-cairo --without-compress-install 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -g -flto -g -flto -s -fuse-ld=gold' CPPFLAGS=-D_FORTIFY_SOURCE=2 LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now)
        windowsys  x
        daemonp    server-running
doom    version    2.0.9
        build      HEAD -> develop 25b240b9b 2020-06-28 01:09:54 -0400
        dir        ~/.doom.d/
system  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /bin/bash
        uname      Linux 5.6.16-1-MANJARO #1 SMP PREEMPT Wed Jun 3 14:26:28 UTC 2020 x86_64
        path       (~/bin ~/.local/bin /usr/local/bin /usr/bin /bin /usr/local/sbin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl ~/.npm-packages/bin ~/.gem/ruby/2.7.0/bin /usr/lib/emacs/27.0.91/x86_64-pc-linux-gnu)
config  envfile    nil
        elc-files  0
        modules    (:completion (company +childframe) (ivy +prescient +childframe +icons) :ui doom doom-dashboard doom-quit hl-todo hydra indent-guides minimap modeline ophints (popup +defaults) pretty-code treemacs vc-gutter vi-tilde-fringe :editor file-templates fold multiple-cursors snippets word-wrap :emacs dired electric (undo +tree) vc :term vterm :checkers syntax :tools docker (eval +overlay) lookup lsp magit pdf rgb :lang data emacs-lisp json (javascript +lsp) markdown (org +noter +pandoc) (php +lsp) (python +lsp) rest (sh +lsp) (web +lsp) yaml :config literate (default +bindings))
        packages   ((phi-search) (phi-search-mc) (crux) (bm) (auto-dim-other-buffers) (rg) (deadgrep) (whole-line-or-region) (mixed-pitch) (org-protocol-capture-html :recipe (:host github :repo alphapapa/org-protocol-capture-html)) (org-download) (org-ql) (org-super-agenda) (org-drill) (org-drill-table) (org-mru-clock) (org-clock-csv) (ox-pandoc) (org-caldav) (org-download) (org-ref) (history) (ox-reveal) (ctrlf) (visual-fill-column))
        unpin      (n/a)
        elpa       (treemacs s pfuture lv hydra ht f dash company bm avy ace-window)
```

</pre></details>

**Comment by tbruckmaier** [comment]
I have made mostly all buffers real now with `(setq doom-unreal-buffer-functions '(minibufferp))` (which was `'(minibufferp doom-special-buffer-p doom-non-file-visiting-buffer-p)` before). Switching tabs works fine now, also `switch-to-next-buffer` recognizes these tabs as expected.

Which leads to the question.. why does doom have this real/unreal buffer concept in the first place? All I could find was the documentation for `doom-real-buffer-p` which states:

```
Returns t if BUFFER-OR-NAME is a 'real' buffer.

A real buffer is a useful buffer; a first class citizen in Doom. Real ones
should get special treatment, because we will be spending most of our time in
them. Unreal ones should be low-profile and easy to cast aside, so we can focus
on real ones.

The exact criteria for a real buffer is:

  1. A non-nil value for the buffer-local value of the doom-real-buffer-p
     variable OR
  2. Any function in doom-real-buffer-functions returns non-nil OR
  3. None of the functions in doom-unreal-buffer-functions must return
     non-nil.
```

What exactly does "first class citizen" and "low-profile" mean? Is it just for skipping certain buffers with buffer change functions or is there more to it? Any enlightenment will be appreciated

**Comment by hlissner** [comment]
TL;DR: use the `:ui tabs` module instead of `tab-bar-mode` directly. It comes with a nice theme by default and gets around this issue by using the correct mechanism for buffer switching (unlike tab-bar-mode).

> When tab-line-mode is active and a *scratch* or any unreal buffer is shown as a tab, clicking it should select that buffer.

There are four problems here.

1. `tab-bar-mode` is using the wrong mechanism for switching to a buffer. A mechanism that consults the frame's buffer-predicate to determine if a buffer is "worthy" of being switched to. The buffer predicate is meant to regulate _untargeting_ buffer switching, not explicit ones. Switching to a particular buffer+tab _should_ be an explicit, targeted switch. `tab-bar-mode` is a young library and likely hasn't been tested thoroughly.

2. The `persp-mode` plugin (from the `:ui workspaces` module) installs a buffer predicate that ignores *unreal* buffers, or buffers that aren't registered to the current workspace.

3. Doom adds its own frame predicate (`doom-buffer-frame-predicate`) which prevents switching to unreal buffers.

4. `*scratch*` buffers are not real buffers.

> Which leads to the question.. why does doom have this real/unreal buffer concept in the first place?

I isolate the useful (real) buffers, so we don't end up in an unreal one unless we explicitly switch to them. It exists to make buffer switching more predictable. Emacs, its plugins, and various processes all produce numerous buffers that we don't want to _accidentally_ switch to, because they are generally unimportant or destructive to be in unintentionally. And there are many operations (mainly to do with killing buffers or switching to next/previous buffer) that could cause us to inadvertently switch to them.

Example: an LSP server produces a `*lsp-log*` buffer. There is no reason to see this buffer unless you're debugging an issue. If it weren't "unreal", you could accidentally land in it by invoking `next-buffer` or when you killing/burying the current buffer. A mild annoyance, but multiply this by the dozens, possibly hundreds of unimportant buffers that could exist in any long lived Emacs session.

Another example: imagine a hidden treemacs buffer. There exist two-key sequences that are capable of deleting entire directory trees in a treemacs buffer (which fortunately ask for confirmation, but...). Imagine this happening within a split second: you bury/kill a file buffer and land in that treemacs buffer, but you already have a new destination in mind. You immediately press some global key binding (maybe the leader key) you _thought_ would invoke something else, but instead invoked one of those "delete this directory tree" keybinds in treemacs. The confirmation prompt pops up. You don't notice why Emacs is blocking. You mash `C-g` (hopefully). Disaster averted, but that was annoying, disruptive and potentially disastrous.

Of course, that kind of danger is unrealistic. Disruption is more likely than damage, but it's not impossible.

This problem is more pronounced for evil users, who have to contend with multiple modes per keymap.

**Comment by fredrikatmainspringenergy** [comment]
> Which leads to the question.. why does doom have this real/unreal buffer concept in the first place?

Is there an option to turn the unreal/real concept off?

Background: I often switch to the `*compilation*` buffer as part of my normal workflow, so for me that "unreal" buffer is essential to have quick access to.

My current workarounds:
1. `C-x` `C-b` and then locate the desired buffer hit Enter
2. `M-x` `switch-to-buffer-other-frame` and then type the desired buffer name (with tab support)

Note: I'm using neither `ui: tabs` nor `tab-bar-mode`

**Comment by tbruckmaier** [comment]
I have moved on from emacs, but I had disabled the real/unreal concept with `(setq doom-unreal-buffer-functions '(minibufferp))`. That makes all buffers "real" and they behave the same. I enjoyed it better like this, but you have to keep hlissners explanations in mind

**Comment by fredrikatmainspringenergy** [comment]
Thanks tbrukmainer, but that fix didn't work for me.
